plugins/hls-class-plugin/src/Ide/Plugin/Class/Types.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE DeriveAnyClass   #-}
+{-# LANGUAGE DeriveGeneric    #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE TypeFamilies     #-}
+{-# LANGUAGE ViewPatterns     #-}
+
+module Ide.Plugin.Class.Types where
+
+import           Control.DeepSeq               (rwhnf)
+import           Control.Monad.Extra           (whenMaybe)
+import           Control.Monad.IO.Class        (liftIO)
+import           Data.Aeson
+import           Data.Maybe                    (catMaybes)
+import qualified Data.Text                     as T
+import           Development.IDE
+import qualified Development.IDE.Core.Shake    as Shake
+import           Development.IDE.GHC.Compat
+import           Development.IDE.Graph.Classes
+import           GHC.Generics
+import           Ide.Plugin.Class.Utils
+import           Ide.Plugin.Config
+import           Ide.Plugin.Properties
+import           Ide.PluginUtils
+import           Ide.Types
+import           Language.LSP.Server
+
+typeLensCommandId :: CommandId
+typeLensCommandId = "classplugin.typelens"
+
+codeActionCommandId :: CommandId
+codeActionCommandId = "classplugin.codeaction"
+
+-- | Default indent size for inserting
+defaultIndent :: Int
+defaultIndent = 2
+
+data AddMinimalMethodsParams = AddMinimalMethodsParams
+    { uri         :: Uri
+    , range       :: Range
+    , methodGroup :: List (T.Text, T.Text)
+    -- ^ (name text, signature text)
+    , withSig     :: Bool
+    }
+    deriving (Show, Eq, Generic, ToJSON, FromJSON)
+
+data GetInstanceBindTypeSigs = GetInstanceBindTypeSigs
+    deriving (Generic, Show, Eq, Ord, Hashable, NFData)
+
+data InstanceBindTypeSig = InstanceBindTypeSig
+    { bindName     :: Name
+    , bindRendered :: T.Text
+    , bindDefSpan  :: Maybe SrcSpan
+    -- ^SrcSpan for the bind definition
+    }
+
+newtype InstanceBindTypeSigsResult =
+    InstanceBindTypeSigsResult [InstanceBindTypeSig]
+
+instance Show InstanceBindTypeSigsResult where
+    show _ = "<InstanceBindTypeSigs>"
+
+instance NFData InstanceBindTypeSigsResult where
+    rnf = rwhnf
+
+type instance RuleResult GetInstanceBindTypeSigs = InstanceBindTypeSigsResult
+
+newtype Log = LogShake Shake.Log deriving Show
+
+instance Pretty Log where
+  pretty = \case
+    LogShake log -> pretty log
+
+data BindInfo = BindInfo
+    { bindSpan     :: SrcSpan
+      -- ^ SrcSpan of the whole binding
+    , bindNameSpan :: SrcSpan
+      -- ^ SrcSpan of the binding name
+    }
+
+rules :: Recorder (WithPriority Log) -> Rules ()
+rules recorder = do
```

Comment:
Could you explain why do you think that it would be better to introduce a new rule?

Reply:
Because rules have cache? I'm not sure about that.